driver-core/src/main/com/mongodb/operation/MixedBulkWriteOperation.java

```diff
@@ -203,64 +245,97 @@ public void call(final AsyncConnection connection, final Throwable t1) {
         });
     }
 
-    private BulkWriteResult executeBatches(final Connection connection, final SessionContext sessionContext) {
-        BulkWriteBatch batch = BulkWriteBatch.createBulkWriteBatch(namespace, connection.getDescription(), ordered,
-                writeConcern, bypassDocumentValidation, writeRequests);
+    private BulkWriteResult executeBulkWriteBatch(final WriteBinding binding, final Connection connection,
+                                                  final BulkWriteBatch originalBatch) {
+        BulkWriteBatch currentBatch = originalBatch;
+        MongoException exception = null;
 
-        while (batch.shouldProcessBatch()) {
-            BsonDocument result = connection.command(namespace.getDatabaseName(), batch.getCommand(), NO_OP_FIELD_NAME_VALIDATOR,
-                    ReadPreference.primary(), batch.getDecoder(), sessionContext, shouldAcknowledge(batch, writeConcern),
-                    batch.getPayload(), batch.getFieldNameValidator());
-            batch.addResult(result);
-            batch = batch.getNextBatch();
+        try {
+            while (currentBatch.shouldProcessBatch()) {
+                BsonDocument result = executeCommand(connection, currentBatch, binding);
+                currentBatch.addResult(result);
+                currentBatch = currentBatch.getNextBatch();
+            }
+        } catch (MongoException e) {
+            exception = e;
+        } finally {
+            connection.release();
         }
 
-        return batch.getResult();
+        if (exception == null) {
+            return currentBatch.getResult();
+        } else if (!currentBatch.shouldProcessBatch() || !originalBatch.getRetryWrites() || !isRetryableException(exception)) {
+            throw exception;
+        } else {
+            return retryExecuteBatches(binding, currentBatch, exception);
+        }
     }
 
-    private BulkWriteResult executeLegacyBatches(final Connection connection) {
-        for (WriteRequest writeRequest : getWriteRequests()) {
-            if (writeRequest.getType() == INSERT) {
-                connection.insert(getNamespace(), isOrdered(), (InsertRequest) writeRequest);
-            } else if (writeRequest.getType() == UPDATE || writeRequest.getType() == REPLACE) {
-                connection.update(getNamespace(), isOrdered(), (UpdateRequest) writeRequest);
-            } else {
-                connection.delete(getNamespace(), isOrdered(), (DeleteRequest) writeRequest);
+    private BulkWriteResult retryExecuteBatches(final WriteBinding binding, final BulkWriteBatch retryBatch,
+                                                final MongoException originalError) {
+        return withReleasableConnection(binding, new CallableWithConnectionAndSource<BulkWriteResult>() {
```

Comment:
It looks like this is violating this part of the spec:

    /* If we cannot select a writable server, do not proceed with retrying and
     * throw the original error. The caller can then infer that an attempt was
     * made and failed. */
    try {
      server = selectServer("writable");
    } catch (Exception ignoredError) {
      throw originalError;
    }

i.e. it's not throwing the original error if server selection fails.

Reply:
Done.